File: 10_short_rate_models_with_stochastic_volatility/cpp/StochasticVolSimulation.cpp
#include "StochasticVolSimulation.hpp"

#include <random>
#include <cmath>

/*
    Simulate correlated short-rate and variance paths.

    Correlated shocks:

        Z_vol = rho Z_rate + sqrt(1-rho^2) Z_independent
*/

StochasticVolSimulationResult simulateStochasticVolPaths(
    const StochasticVolModel& model,
    double r0,
    double v0,
    double maturity,
    int nSteps,
    int nPaths,
    unsigned int seed
){

    double dt =
        maturity
        /
        static_cast<double>(nSteps);

    StochasticVolSimulationResult result;

    result.times.resize(
        nSteps + 1
    );

    for(int i = 0; i <= nSteps; ++i){

        result.times[i] =
            i * dt;
    }

    result.ratePaths =
        std::vector<std::vector<double>>(
            nPaths,
            std::vector<double>(
                nSteps + 1,
                0.0
            )
        );

    result.variancePaths =
        std::vector<std::vector<double>>(
            nPaths,
            std::vector<double>(
                nSteps + 1,
                0.0
            )
        );

    for(int p = 0; p < nPaths; ++p){

        result.ratePaths[p][0] = r0;

        result.variancePaths[p][0] = v0;
    }

    std::mt19937 generator(seed);

    std::normal_distribution<double> normal(
        0.0,
        1.0
    );

    for(int step = 0; step < nSteps; ++step){

        double t =
            result.times[step];

        for(int p = 0; p < nPaths; ++p){

            double zRate =
                normal(generator);

            double zIndependent =
                normal(generator);

            double zVol =
                model.rho * zRate
                +
                std::sqrt(
                    1.0
                    -
                    model.rho
                    *
                    model.rho
                )
                *
                zIndependent;

            double rNext;
            double vNext;

            model.evolve(
                t,
                result.ratePaths[p][step],
                result.variancePaths[p][step],
                dt,
                zRate,
                zVol,
                rNext,
                vNext
            );

            result.ratePaths[p][step + 1] =
                rNext;

            result.variancePaths[p][step + 1] =
                vNext;
        }
    }

    return result;
}

/*
    Discount factor from startIndex to endIndex:

        exp(- integral r_t dt)
*/

double pathDiscountFactor(
    const std::vector<double>& times,
    const std::vector<double>& ratePath,
    int startIndex,
    int endIndex
){

    double dt =
        times[1]
        -
        times[0];

    double integral =
        0.0;

    for(int i = startIndex; i < endIndex; ++i){

        integral +=
            ratePath[i]
            *
            dt;
    }

    return
        std::exp(
            -integral
        );
}

/*
    Discount factor from 0 to endIndex.
*/

double pathDiscountFactorFromZero(
    const std::vector<double>& times,
    const std::vector<double>& ratePath,
    int endIndex
){

    return
        pathDiscountFactor(
            times,
            ratePath,
            0,
            endIndex
        );
}

File: 10_short_rate_models_with_stochastic_volatility/python/stochastic_vol_simulation.py
"""
stochastic_vol_simulation.py

Monte Carlo simulation for the stochastic-volatility short-rate model.
"""

import numpy as np


def simulate_stochastic_vol_paths(
    model,
    r0,
    v0,
    maturity,
    n_steps,
    n_paths,
    seed=42
):
    """
    Simulate correlated short-rate and variance paths.
    """

    np.random.seed(seed)

    dt = maturity / n_steps

    times = np.linspace(
        0.0,
        maturity,
        n_steps + 1
    )

    rate_paths = np.zeros(
        (n_paths, n_steps + 1)
    )

    variance_paths = np.zeros(
        (n_paths, n_steps + 1)
    )

    rate_paths[:, 0] = r0
    variance_paths[:, 0] = v0

    for step in range(n_steps):

        t = times[step]

        z_rate = np.random.normal(
            size=n_paths
        )

        z_independent = np.random.normal(
            size=n_paths
        )

        z_vol = (
            model.rho * z_rate
            +
            np.sqrt(1.0 - model.rho ** 2)
            *
            z_independent
        )

        for path in range(n_paths):

            r_next, v_next = model.evolve(
                t=t,
                r=rate_paths[path, step],
                v=variance_paths[path, step],
                dt=dt,
                z_rate=z_rate[path],
                z_vol=z_vol[path]
            )

            rate_paths[path, step + 1] = r_next
            variance_paths[path, step + 1] = v_next

    return times, rate_paths, variance_paths


def path_discount_factors(
    times,
    rate_paths,
    end_index=None
):
    """
    Compute pathwise discount factors:

        D(0,T) = exp(- integral_0^T r_t dt)
    """

    if end_index is None:
        end_index = rate_paths.shape[1] - 1

    dt = times[1] - times[0]

    integrals = np.sum(
        rate_paths[:, :end_index],
        axis=1
    ) * dt

    return np.exp(
        -integrals
    )


def forward_discount_factors(
    times,
    rate_paths,
    start_index,
    end_index
):
    """
    Compute pathwise discount factors from start_index to end_index:

        D(t,T) = exp(- integral_t^T r_s ds)
    """

    dt = times[1] - times[0]

    integrals = np.sum(
        rate_paths[:, start_index:end_index],
        axis=1
    ) * dt

    return np.exp(
        -integrals
    )